x/cronos/keeper/precompiles/ica.go

```diff
@@ -0,0 +1,168 @@
+	cronoseventstypes "github.com/crypto-org-chain/cronos/v2/x/cronos/events/types"
+	icaauthkeeper "github.com/crypto-org-chain/cronos/v2/x/icaauth/keeper"
+	"github.com/crypto-org-chain/cronos/v2/x/icaauth/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+// TODO: Replace this const with adjusted gas cost corresponding to input when executing precompile contract.
+const ICAContractRequiredGas = 10000
+
+var (
+	icaABI             abi.ABI
+	icaContractAddress = common.BytesToAddress([]byte{102})
+)
+
+func init() {
+	if err := icaABI.UnmarshalJSON([]byte(ica.ICAModuleMetaData.ABI)); err != nil {
+		panic(err)
+	}
+}
+
+type IcaContract struct {
+	BaseContract
+
+	cdc           codec.Codec
+	icaauthKeeper *icaauthkeeper.Keeper
+}
+
+func NewIcaContract(icaauthKeeper *icaauthkeeper.Keeper, cdc codec.Codec) vm.PrecompiledContract {
+	return &IcaContract{
+		BaseContract:  NewBaseContract(icaContractAddress),
+		cdc:           cdc,
+		icaauthKeeper: icaauthKeeper,
+	}
+}
+
+func (ic *IcaContract) Address() common.Address {
+	return icaContractAddress
+}
+
+// RequiredGas calculates the contract gas use
+func (ic *IcaContract) RequiredGas(input []byte) uint64 {
+	return ICAContractRequiredGas
+}
+
+func (ic *IcaContract) IsStateful() bool {
+	return true
+}
+
+func (ic *IcaContract) Run(evm *vm.EVM, contract *vm.Contract, readonly bool) ([]byte, error) {
+	// parse input
+	methodID := contract.Input[:4]
+	method, err := icaABI.MethodById(methodID)
+	if err != nil {
+		return nil, err
+	}
+
+	stateDB := evm.StateDB.(ExtStateDB)
+	precompileAddr := ic.Address()
+	caller := contract.CallerAddress
+	owner := sdk.AccAddress(caller.Bytes()).String()
+	converter := cronosevents.IcaConvertEvent
+	var execErr error
+	switch method.Name {
+	case "registerAccount":
+		if readonly {
+			return nil, errors.New("the method is not readonly")
+		}
+		args, err := method.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		version := args[1].(string)
+		execErr = stateDB.ExecuteNativeAction(precompileAddr, converter, func(ctx sdk.Context) error {
+			_, err := ic.icaauthKeeper.RegisterAccount(ctx, &types.MsgRegisterAccount{
+				Owner:        owner,
+				ConnectionId: connectionID,
+				Version:      version,
+			})
+			return err
+		})
+		if execErr != nil {
+			return nil, execErr
+		}
+		return method.Outputs.Pack(true)
+	case "queryAccount":
+		args, err := method.Inputs.Unpack(contract.Input[4:])
+		if err != nil {
+			return nil, errors.New("fail to unpack input arguments")
+		}
+		connectionID := args[0].(string)
+		account := args[1].(common.Address)
+		owner := sdk.AccAddress(account.Bytes()).String()
+		icaAddress := ""
+		execErr = stateDB.ExecuteNativeAction(precompileAddr, nil, func(ctx sdk.Context) error {
```

Comment:
Queries can be done directly, more efficient than `ExecuteNativeAction`, I can add a primitive in ethermint first.

Reply:
here we are: https://github.com/crypto-org-chain/ethermint/pull/359